### ISSUE TITLE
Feat: Error handling on the Backend to report human errors

### ISSUE THREAD

**Issue by Emyrk** [coder/coder#1983]
## What is your suggestion?

Ensure backend errors returned to the UI are human readable. Currently we elevate `err.Error()` a lot, which is not a good UI experience.

```golang
httpapi.Write(rw, http.StatusForbidden, httpapi.Response{
     Message: "Human message here",
     Internal: err.Error(),
})
```

## Why do you want this feature?

So the backend can send appropriate errors over to the frontend.

## Are there any workarounds to get this functionality today?

No

## Are you interested in submitting a PR for this?

Yes

**Comment by presleyp** [comment]
(notes from our discussion) The frontend needs to be able to distinguish:
1. general, user-friendly error messages (`message`). They will be shown at the top/bottom of a form (not associated with a specific form field), on a non-form page, or in a toast.
2. form field-specific, user-friendly error messages (`errors`). They will be shown next to the associated form field.
3. technical error info (`internal`). This will not be displayed on the page but will be accessible in the inspector for debugging purposes.

**Comment by presleyp** [comment]
Maybe we can have @khorne3 review this PR too, to make sure things are capitalized right and all that.

**Comment by kylecarbs** [comment]
What would the technical error info indicate? In v1, we displayed "A database error occurred.", which most customers that have encountered it complained about. Since our primary user is the developer, it might be best to show a technical error to start, and simplify it if we get complaints.

**Comment by kylecarbs** [comment]
I'm entirely on board with displaying human errors, but I think our technical errors should be human-readable too! We're all human! 🤖🤖🤖

**Comment by Emyrk** [comment]
@kylecarbs The `message` is in the toast pop up. `A database error occurred` is sufficient for that.

The actual error could be `duplicate key value violates unique constraint some_key`. I guess the question is, what do we want to display. This is a postgres error, which isn't too bad. But some errors (especially network errors of just `EOF`) are quite vague and not helpful to a user.

![Screenshot from 2022-06-02 11-11-26](https://user-images.githubusercontent.com/5446298/171675838-f7931e9a-00de-480e-95d1-a818f6b91883.png)

(I mean't to put insert vs fetch, but same gist)

![Screenshot from 2022-06-02 11-11-10](https://user-images.githubusercontent.com/5446298/171675821-60aeb389-5a38-4a23-b8ab-f72103457e2a.png)

----

The idea of the `internal` field was to always include the PG error for debugging, but don't show it to the user, as they can't do anything with it.

**Comment by Emyrk** [comment]
I can live with the more detailed later in the toast if we are all ok with it.

**Comment by kylecarbs** [comment]
A user can send the error to an admin, who can do something. With "A database error occurred," there's nothing they can do. They can just tell their admin that Coder is broken.

**Comment by Emyrk** [comment]
That error is a bad error though. It's the common bad example when we get an internal error that shouldn't ever happen and should trigger a log.

I'm not opposed though. See https://github.com/coder/coder/pull/1994 for some examples. I think this is fine for now, to not use `Internal` if we are ok showing that stuff

**Comment by kylecarbs** [comment]
I'm not sure where to draw the line on what is a friendly error versus an unhelpful one. On one hand, it's friendlier to say "Failed to create user because the database is disconnected" but on another, it's much more actionable to say "Failed to create user: the database has too many open connections". A user would presumably send this to an admin, who would immediately know what they need to fix.

I'd prefer for us to print the raw errors in the message for now (like exists today), and if we find it too janky or difficult to format in the UI we discuss a method for drawing the distinction.

**Comment by kylecarbs** [comment]
Confusing errors are different than humanizing errors.

We have equivalent examples of where customers encounter "failed to dial workspace" without any action or insight into the cause. I'm trying to avoid the situation where customers encounter a human error and are told to open dev-tools only to discover they already missed the request. Our answer has been "I guess wait for it to happen again..." historically, which is worse than displaying an overly verbose and more confusing error.

GitHub has only needed one field to this point, but that's just one API and might not be entirely representative: https://docs.github.com/en/rest/overview/resources-in-the-rest-api#client-errors

**Comment by misskniss** [comment]
Humanize the errors. We want them useful and will continue to into the future. These are the details that instill high confidence _despite_ their appearance as errors in the first place. 

I like @kylecarbs point that we also the human consumers of the errors so they should be meaningful to developers as well. 

I like the user facing errors that do not give away details about internals anymore than necessary and provide some guidance about what to do where reasonable. 

Does the user have clear enough information about what went wrong so that they can: 
1. Fix the error if it is caused by them like  "A User with this email already exists", etc
2. Share the error with support and we will know where to start looking to fix it:  "User Save Failed: too many connections please try again later"

**Comment by Emyrk** [comment]
`A User with this email already exists` is an example when we know the error and can confidently just replace it with a message. It's the case where things are ambigious. The most common line of code with is Database internal errors. 

These errors can be "too many connections open", some column constraint failure, some timeout, etc, etc. These errors require the details to be useful to debug, but to the user on the FE, the error is just as unactionable as "A database error occurred".

Some other examples:
- External packages. When we call external packages, we have no control over the internal error portion. The human message is likely to be vague. 
- Internal packages. Golang errors are inherently chained and not formatted as we'd like for the UI. Calling anything in the peer package for example is likely to be technical. 


--------

**A**
> ![Screenshot from 2022-06-02 11-11-26](https://user-images.githubusercontent.com/5446298/171675838-f7931e9a-00de-480e-95d1-a818f6b91883.png)

**B**
> ![Screenshot from 2022-06-02 11-11-10](https://user-images.githubusercontent.com/5446298/171675821-60aeb389-5a38-4a23-b8ab-f72103457e2a.png)


I think we just need to say A or B. If we choose **A**, I will still return these errors in the response payload (assuming no confidential info or security concern), but the UI will not display it.

@misskniss @ammario It sounds like you lean to **A**? Meaning the case where the BE can't really determine what the error is, we return a vague, but human error on the UI. Is this correct?

**Comment by presleyp** [comment]
@f0ssel has a good idea about having the nitty gritty error collapsed but accessible, but that requires it staying for a while (hard with toasts)

**Comment by Emyrk** [comment]
> @f0ssel has a good idea about having the nitty gritty error collapsed but accessible, but that requires it staying for a while (hard with toasts)

I was thinking this too, but in that case I'd rather separate it by `internal` and `message` so the FE team knows which part is which (for the collapse)

**Comment by presleyp** [comment]
Yeah, we could `console.error` the `internal` error message as a stopgap while we develop collapsible error message components.

**Comment by Emyrk** [comment]
@ammario I think that encapsulates what Presley mentioned of some collapsible field.

**Comment by kylecarbs** [comment]
I like the collapsible field! Maybe we do that and `console.error`, just in case they missed the toast.

**Comment by misskniss** [comment]
I like the flag/env vars for the expanded UI option suggested by @presleyp and @ammario . How much does that change the complexity for you @Emyrk and @presleyp? Does that make it unreasonably complex for the MVP? If not, I am yes on option A+ with quick logging exceptions follow up.

**Comment by kylecarbs** [comment]
Just to be clear, I'm good with `A` as long as the context is captured in those other places.

**Comment by misskniss** [comment]
🎉

**Comment by Emyrk** [comment]
> I like the flag/env vars for the expanded UI option suggested by @presleyp and @ammario . How much does that change the complexity for you @Emyrk and @presleyp? Does that make it unreasonably complex for the MVP? If not, I am yes on option A+ with quick logging exceptions follow up.

If we break it into 2 fields `message` and `internal`, it gives us a ton of flexibility how to handle it. (Collapse, env var flag, console log, etc)

**Comment by ketang** [comment]
There should be a clear distinction between user level errors and system level errors.  "This email address is already in use" I'd characterize as a user level error.  We need to make those as comprehensible as we realistically can.  System level errors like "too many database connections" don't need to be made friendly or local.

**Comment by ammario** [comment]
> One way to distinguish is that typically user-level error is something we explicitly check for while a system level error is something that happens to us.

There's also the classic 4xx vs. 5xx for the FE.

**Comment by presleyp** [comment]
> I think human readability needs to be a responsibility shared between back end and front end; I think it's unreasonable to ask the front end to know every potential error the back end could trigger. This becomes even more true when we tackle internationalization and localization.

I'm confused about whether you're saying the backend needs more or less of the responsibility. This ticket is about giving the backend all of the responsibility; the frontend will just display the error message it got. That addresses the "frontend knowing everything" problem, but I'm not sure it addresses internationalization (I've only done it on the frontend but maybe it's commonly done on the backend too).

**Comment by ketang** [comment]
Well, we decide what the status code is on the back end.  I'd say that if it's something that's reasonably (*hand wave*) checkable before attempting execution, we should treat it as a user level error that needs friendly language.

**Comment by presleyp** [comment]
> There should be a clear distinction between user level errors and system level errors.

@ketang what do you mean by distinction? Should the frontend treat them differently?

**Comment by Emyrk** [comment]
@presleyp when we get to internationalization, if it becomes a requirement of the BE... I'm sure we can come up with something.

**Comment by Emyrk** [comment]
Merged: https://github.com/coder/coder/pull/1994
